testlapack/dgeev.go

```diff
@@ -0,0 +1,693 @@
+			evWant: Grcar{N: 50, K: 30}.Eigenvalues(),
+		},
+		{
+			a:      Grcar{N: 150, K: 2}.Matrix(),
+			evWant: Grcar{N: 150, K: 2}.Eigenvalues(),
+		},
+		{
+			a:      Grcar{N: 150, K: 148}.Matrix(),
+			evWant: Grcar{N: 150, K: 148}.Eigenvalues(),
+		},
+
+		{
+			a:      Hanowa{N: 6, Alpha: 17}.Matrix(),
+			evWant: Hanowa{N: 6, Alpha: 17}.Eigenvalues(),
+		},
+		{
+			a:      Hanowa{N: 50, Alpha: -1}.Matrix(),
+			evWant: Hanowa{N: 50, Alpha: -1}.Eigenvalues(),
+		},
+		{
+			a:      Hanowa{N: 100, Alpha: -1}.Matrix(),
+			evWant: Hanowa{N: 100, Alpha: -1}.Eigenvalues(),
+		},
+
+		{
+			a:      Lesp(2).Matrix(),
+			evWant: Lesp(2).Eigenvalues(),
+		},
+		{
+			a:      Lesp(3).Matrix(),
+			evWant: Lesp(3).Eigenvalues(),
+		},
+		{
+			a:      Lesp(4).Matrix(),
+			evWant: Lesp(4).Eigenvalues(),
+		},
+		{
+			a:      Lesp(5).Matrix(),
+			evWant: Lesp(5).Eigenvalues(),
+		},
+		{
+			a:      Lesp(10).Matrix(),
+			evWant: Lesp(10).Eigenvalues(),
+		},
+		{
+			a:      Lesp(15).Matrix(),
+			evWant: Lesp(15).Eigenvalues(),
+		},
+		{
+			a:      Lesp(30).Matrix(),
+			evWant: Lesp(30).Eigenvalues(),
+		},
+		{
+			a:      Lesp(50).Matrix(),
+			evWant: Lesp(50).Eigenvalues(),
+			tol:    1e-12,
+		},
+		{
+			a:      Lesp(101).Matrix(),
+			evWant: Lesp(101).Eigenvalues(),
+			tol:    1e-12,
+		},
+		{
+			a:      Lesp(150).Matrix(),
+			evWant: Lesp(150).Eigenvalues(),
+			tol:    1e-12,
+		},
+
+		{
+			a:      Rutis{}.Matrix(),
+			evWant: Rutis{}.Eigenvalues(),
+		},
+
+		{
+			a:      Tris{N: 74, X: 1, Y: -2, Z: 1}.Matrix(),
+			evWant: Tris{N: 74, X: 1, Y: -2, Z: 1}.Eigenvalues(),
+		},
+		{
+			a:      Tris{N: 74, X: 1, Y: 2, Z: -3}.Matrix(),
+			evWant: Tris{N: 74, X: 1, Y: 2, Z: -3}.Eigenvalues(),
+		},
+		{
+			a:      Tris{N: 75, X: 1, Y: 2, Z: -3}.Matrix(),
+			evWant: Tris{N: 75, X: 1, Y: 2, Z: -3}.Eigenvalues(),
+		},
+
+		{
+			a:      Wilk4{}.Matrix(),
+			evWant: Wilk4{}.Eigenvalues(),
+		},
+		{
+			a:      Wilk12{}.Matrix(),
+			evWant: Wilk12{}.Eigenvalues(),
+			tol:    1e-8,
+		},
+		{
+			a:      Wilk20(0).Matrix(),
+			evWant: Wilk20(0).Eigenvalues(),
+		},
+		{
+			a:      Wilk20(1e-10).Matrix(),
+			evWant: Wilk20(1e-10).Eigenvalues(),
+			tol:    1e-12,
+		},
+
+		{
+			a:      Zero(1).Matrix(),
+			evWant: Zero(1).Eigenvalues(),
+		},
+		{
+			a:      Zero(10).Matrix(),
+			evWant: Zero(10).Eigenvalues(),
+		},
+		{
+			a:      Zero(50).Matrix(),
+			evWant: Zero(50).Eigenvalues(),
+		},
+		{
+			a:      Zero(100).Matrix(),
+			evWant: Zero(100).Eigenvalues(),
+		},
+	} {
+		for _, jobvl := range []lapack.JobLeftEV{lapack.ComputeLeftEV, lapack.None} {
+			for _, jobvr := range []lapack.JobRightEV{lapack.ComputeRightEV, lapack.None} {
+				for _, extra := range []int{0, 11} {
+					for _, optwork := range []bool{false, true} {
+						testDgeev(t, impl, strconv.Itoa(i), test, jobvl, jobvr, extra, optwork)
+					}
+				}
+			}
+		}
+	}
+
+	// for _, n := range []int{10, 30, 50} {
```

Comment:
I expected that commenting this out would trigger questions which in turn would provide a place for explanation and discussion.

It depends. The test matrices above are nice but many of them have too simple structure that will not trigger the inner parts of the algorithm. I wanted to have a couple of tests on (random) matrices with known eigenvalues that are a mix of real and complex conjugate pair numbers. Starting with the real Schur decomposition theorem that for every nxn matrix there exists an orthogonal matrix Q such that 'Q^T A Q = T' where T is in real Schur form (quasi upper triangular with 1x1 and 2x2 blocks on the diagonal), I generate a random T and a random Q and multiply them to get my random A with known eigenvalues. It works but the computed eigenvalues are rather inaccurate, I would have to use tolerance 1e-4 or even 1e-2 (it gets worse with increasing n and affects both real and complex eigenvalues). That's suspiciously loose and that's why I commented it out.

Possible places to look at:
- I'm not using the real Schur decomposition theorem (for every matrix there exists ...), rather "for any T in real Schur form and Q orthogonal, `Q T Q^T` has the same eigenvalues as T". I guess it's fine but I don't have a proof that it holds. I have seen using this approach to generate random symmetric matrices with given eigenvalues, but not for general non-symmetric matrices.
- There might be something wrong with `reflector` and `applyReflector` functions although it does seem that the matrices are indeed orthogonal.
- The randomness creates bad matrices so that `Q T Q^T` has slightly different eigenvalues than T. I suspect this is the reason.

I'll try to think about it further but any comments are appreciated anyway.